curly/data/man/curly.1.md
% Curly(1) Compilers | A minimal Curly Calculus compiler
% Marc Coiffier

SYNOPSIS
========

**Usage:** curly ARGUMENT...

    ARGUMENT = INPUT | TARGET | SCRIPT_NAME | "+"FLAG
    INPUT = ("-M"|"--mount") MODPATH"="MODULE
    MODPATH = [NAME MODPATH]
    MODULE = "source:"SOURCEPATH[ CACHEPATH]
           | "library:"LIBID | "library:@"LIBRARY_FILE
           | "builtins"
    TARGET = "-h" | "--help" | "-v" | "--version"
           | "--instance" INSTANCE | ":"INSTANCE
           | "--banner"["+"] BANNER_TEXT
           | "--at" [SERVER]"/"INSTANCE
           | "--prelude"["+"] COMMAND
           | "-i" | "--interactive"
           | ("-e"|"--execute") COMMAND | "%"COMMAND
           | ("-r"|"--run") CMDFILE
           | ("-s"|"--serve") SERVER_TYPE | ("-l"|"--list") SERVER_TYPE
           | ("-d"|"--dump") SRCLIBNAME
           | ("-t"|"--translate") FILEPATH["@"SYSTEM]"="MODPATH
    SERVER_TYPE = "libraries" | "instances"
    SERVER = "_" | IP_ADDRESS | DNS_NAME
    
    LIB = "@"LIBID
        | FILEPATH".cyl"
        | FILEPATH".cy"

    CONDSPEC = "+"ARGFLAGS CONDSPEC
             | ARGCLAUSE
    ARGFLAGS = FLAG[","ARGFLAGS]
    ARGCLAUSE = "mount" MODPATH "=" MODULE
              | ("target"|"-") TGTSPEC
              | "%" COMMAND
              | ">" ECHO_STRING
              | "include" MODPATH "=" SCRIPT_NAME
    TGTSPEC = "help" | "version"
            | "banner"["+"] BANNER_TEXT
            | "instance" INSTANCE
            | "at" [SERVER]"/"INSTANCE
            | "prelude"["+"] COMMAND
            | "interactive" | "execute" COMMAND
            | "run" CMDFILE
            | "serve" SERVER_TYPE | "list" SERVER_TYPE
            | "dump" SRCLIBNAME
            | "translate" FILEPATH[@SYSTEM] "=" MODPATH

DESCRIPTION
===========

Curly is a humble compiler from Curly Calculus (a derivative of
lambda-calculus) to machine code. It can find, fetch, parse, run,
compile and/or distribute Curly code, automatically resolving
dependencies as it does so.

This command works in a simple way. It constructs an evaluation
context from the INPUTs, then produces each TARGET within that
context. INPUTs can come from various places, and TARGETs can produce
various results.

INPUTS
======

When Curly encounters an import statement in a source file or an
interactive session, it zips that import with the context constructed
from the INPUTs that were given as arguments.

An INPUT can either be a verbatim mount specification, as seen above,
or a Curly script. If it is the latter, that script file should
contain a list of conditional argument specifications (CONDSPEC), one
per line. Each argument can be flagged by one or many arbitrary flag
names, any of which can be specified on the command-line (with +FLAG)
to enable that argument, thus enabling some sort of modal
compilation. Relative paths in an argument shall be treated relative
to the directory containing the script instead of the working
directory.

The default context is constructed by starting with an empty module,
and mounting each module at its path for each mount
specification. Modules can be defined from source files and
directories, library files or library IDs, which are simply sha256
checksums of their corresponding library files.

Library IDs are used by the compiler to unambiguously locate libraries
on the local filesystem or on distant repositories.

Whenever Curly needs a library, it caches it in the directory
specified by the environment variable CURLY_LIBCACHE (or
$HOME/.curly/libraries if unset). If a library can't be found in the
cache, the environment variable CURLY_PATH contains a list of
comma-separated Curly repositories that can be queried for that
library. If none of the repositories know the library, only then does
Curly fail to resolve the dependency.

There are two kinds of Curly repositories : native Curly repositories,
of the form @host[:port] (see the --serve option for more information
about curly repositories); and backend repositories, of the form
*\<backend-name\>(:\<arg\>)...*. A backend repository is an executable
in the '/usr/share/curly/backend' directory, which takes a library ID
as its first argument, and outputs the corresponding library's
contents, if it can be found. This behaviour makes it easy to write
your own custom backends to retrieve libraries in non-standard
ways. Three backends are already provided in this bundle : a backend
to retrieve libraries from HTTP repositories, with
'http://server.example/prefix'; a backend for local directories, with
'local:directory'; and a backend for the Curly version control system,
with 'curly-vc://user@host[:port]/branch' (more on Curly version
control can be found at )

HTTP and directory repositories must be structured in a certain way
for Curly to be able to query them (Curly repositories already offer
the expected functionality, so nothing is to be done for them) :

  * Any available library of hash H should be found in the file H.cyl.
  * There must be a file named "libraries" containing a list of
    library identifiers, one per line, that are available in this
    repository

You'll note that repositories may not be specified as flags on the
command-line. The reason for that is simple : if Curly succeeds in
producing an output, that output will be the same regardless of the
available repositories. Since repositories do not affect the result of
compilation, they should be part of the the environment rather than
the command-line.

TARGETS
=======

Once its inputs are properly positioned, Curly can produce a variety
of outputs and effects :

-h,--help,-v,--version

:   The usual help and version flags. The --help flag inhibits all
others, in order to allow you to see what Curly actually understood of
the options you gave it.

-i,--interactive

:   Launches an interactive session. You can import anything mounted
from the INPUTs in this session. If a previous --at option was set to a
nonempty value, launches a distant interactive session on that server
instead.

-e, --execute, %

:   Runs its parameter, which can be any valid interactive
expression. The third form can be used as a shortcut, like *./context
%clean*. Like -i, the expression is sent to the given server set from
a previous --at option.

-d,--dump

:   Shows the contents of the given Curly source or library file. Can
also show the contents of a library from its ID, with the special
syntax **@<library-id>**

-s,--serve libraries

:   Launches a Curly library server over the port specified by the
CURLY_PORT environment variable, or 25465 if the latter is
unset. Curly clients can use this server as a repository by adding
`@host[:port]` to their CURLY_PATH. If the port is omitted, Curly
connects to the same port it would listen on.

    This proxy forwards requests to its own CURLY_PATH, caching them
for future requests to avoid repetitious downloads.

    Many library servers can borrow the same port number. Each
additional library server connects to the process holding the socket
and serves libraries that the latter can forward to its other clients.

-s,--serve instances

:   Launches an interactive server for the current instance through the
port specified by CURLY_PORT (or 25465 by default). Clients can
connect to this server by specifying the --at option before launching
an interactive session or running a batch command or a command file.

    For example, if you want to make a given context accessible by the
name 'myProject', you can run **curly context --instance myProject -si**
for a server, to which clients can connect by running **curly
--at \<server-address\>/myProject -i**. The alternate syntax **curly
@\<server-address\>/myProject -i** is also recognized. If the server
address is omitted, it defaults to 127.0.0.1.

    Like with library servers, several instance servers can use the
same port, with the caveat that the main server only acts as a port
mapper for the instance servers. Clients connect to the main port to
ask for the port number of the instance they need, and servers ask it
to allocate a free port number number for them to listen on.

    The port mapper's allocation strategy is simple : allocate the next
port above CURLY_PORT that isn't already used by another instance.
Thus, if a machine hosting instance servers is behind a firewall, that
firewall should leave several ports numbers above CURLY_PORT open if
clients are to be able to connect to those instances.

-l,--list libraries

:   Lists all available libraries from the current CURLY_PATH.

-l,--list instances

:   Lists all instances that are available on the server previously
specified with the --at option, or localhost if no server was
specified.

-t,--translate

:   Translates a Curly function into an executable file, or some other
kind of representation of that function. In the future, Curly will be
able to generate Javascript, Verilog and other intermediate languages
for which an ABI cannot readily be determined, as well as executables
for alternative systems like Windows or MacOS.

--prelude, --prelude+

:   Sets or adds to the prelude. The prelude is a list of commands that
are run before each local session. It is especially useful when
serving instances, as it allows a default context to be defined for
all clients.

--banner

:   Uses the contents of a file as the initial text that is displayed
before each interactive session is started. When set before serving
instances, this sets the banner for each client.

--at

:   Sets the server to connect to when running interactive sessions,
commands or files. If the server is empty, those sessions are hosted
in the local context instead.

--instance

: Changes the name of the current instance. This name is used in the
prompt of client interactive sessions and as a handle for the instance
server, if one is launched. Curly also accepts the alternate syntax
**:INSTANCE** for this flag.

METADATA
========

Since version 0.50.3, Curly libraries can contain arbitrary metadata,
in the form of a hierarchy of strings (sort of like a simplified
DOM). This metadata can be leveraged by the compiler to search its
repositories for libraries based on their external attributes.

For example, instead of manually locating the library 'stdlib' in your
repositories and mounting its hash, you can now let Curly do the
searching, by writing "curly --mount stdlib=package:'{= {$ name}
stdlib}'". Queries are written in a small language that provides basic
pattern-matching and general numeric comparison, as well as the
ability to test the presence of a metadata field.

Although they can be used for locating packages, queries can also
format the results of a library listing. With every command that
displays a library index, you can customize how the metadata is
displayed by specifying a query to run. For example, here is how you
would list libraries with their names and version numbers :

    curly --list 'libraries {$ name}-{$ version}:\ {$ synopsis}'

Since the format is also a query, you can use it to restrict the
number of libraries that are printed. For example, the next command
only shows different versions of 'stdlib', otherwise using the same
format as before :

    curly --list 'libraries {= {$ name} stdlib}-{$ version}:\ {$ synopsis}'

SECURITY
========

Since Curly can host interactive sessions accessible from other
computers, there is a potential security risk if we let those sessions
run arbitrary code from unauthenticated clients.

That is why, as of version 0.44, Curly can authenticate clients before
allowing them access to an instance. You can still host
unauthenticated instances if you wish, but those only grant readonly
access to their clients.

As with everything else, Curly tries to make security as simple to setup
and use as possible. Case in point, here are the steps required to
host a secure instance that allows all access only to a single client :

  * Generate a private key on the client : **cy %'key gen \<key-name\>'**
  * Export a claim for that key : **cy %'key export \<key-name\>'**
  * Import that claim on the server : **cy %'key import \<key-name\> #\<key-export\>'**
  * Grant access to the target instance for that claim **cy --instance=\<instance\> %'key grant almighty \<key-name\>'**
  * Launch the daemon on the server **cy --instance=\<instance\> -si**

The client can now connect to the instance in the usual ways, as long
as it holds an authorized key. If the client and server agree on
several keys, the one with the highest access level is chosen.

VERSION CONTROL
===============

As of version 0.50.2, Curly comes with a builtin interface to its own
distributed version control system, to help you track and retrieve the
results of your work.

Curly store this information on a DHT for redundancy and cache
behaviour, as well as towards an ideal of global code sharing. If you
use the default Curly server, everything you commit will be made
publicly available, from source to libraries. However, if you prefer
working privately, you can easily host your own local version control
server, by installing the 'curly-dht' package from Hackage and running
'curly-dht \<server-port\>'.

In order to prevent fake source information from being spread through
the DHT, every piece of data is either hash-indexed (using a
cryptographically secure hash function), or signed by a public key
that can be retrieved through the value's index. Thus, every peer can
verify every piece of information it is offered before propagating it.

Under other aspects, version control in Curly works similarly to other
such systems. Every identity can store its own list of branches, which
are each a reference to the tip of a commit chain. Every branch
describes an incrementally-defined library index, which can be used to
locate libraries based on the metadata they export (which can
include a package's name, version and author, among other things).

Since version 0.50.3, Curly also provides a 'checkout' function, that
recreates a full project directory from a library of your choice. This
directory contains the source and libraries for all the dependencies
of the original library, as well as a context file that encapsulates
those dependencies. Together, those files provide a working
environment in which you are free to make changes as you see fit (as
opposed to using raw libraries, which do not allow modifications).


curly/exe/Curly.hs
{-# LANGUAGE CPP, ExistentialQuantification, ViewPatterns #-}
module Main(
  -- * The main Curly function
  main,

  -- * Functions for running your own Curly instances
  initCurly,runTargets,
  ) where

import Control.Concurrent (forkIO,forkFinally)
import Control.Concurrent.MVar (putMVar,takeMVar,newEmptyMVar,MVar)
import Control.Exception (AsyncException(..),Exception(..),bracket_)
import Curly.Core
import Curly.Core.Library
import Curly.Core.Parser
import Curly.Core.Peers
import Curly.Core.Security
import Curly.Core.VCS
import Curly.Core.VCS.Diff (patch)
import Curly.Session.Commands.Common (getDataFileName_ref)
import Curly.Session
import Curly.System
import Curly.System.Base
import Curly.UI
import Curly.UI.Options hiding (nbsp,spc)
import Data.IORef
import GHC.Conc (threadDelay)
import GHC.IO.Encoding (utf8,setLocaleEncoding)
import IO.Filesystem
import IO.Network.Socket
import IO.Time (currentTime)
import Language.Format
import Paths_curly
import System.Directory (doesFileExist)
import System.Environment (getArgs)
import System.IO (withFile,IOMode(..))
import System.Process (readProcess)

main :: IO ()
main = cli "curly" $ do
  initCurly
  
  args <- parseCurlyArgs <$> getArgs
  additional <- liftA2 (+)
                (existingFiles [curlyUserDir</>"default.curly"])
                (if any (has t'1) args then pure [] else existingFiles [".curly"])
  
  let fullArgs = map Left additional + args

  let ?commandLineScripts = [s | Left s <- fullArgs]
  withCurlyConfig fullArgs $ withCurlyPlex ?curlyConfig $ do
    let uninhibited = not ( any (has t'Help) (?curlyPlex^.targets)
                            || all (has t'setting) (?curlyPlex^.targets) )
        tgts = try [Help] (guard uninhibited >> ?curlyPlex^.targets)
    runTargets tgts

existingFiles :: [String] -> IO [String]
existingFiles fs = do
  exs <- traverse doesFileExist fs
  return (fold (zipWith (\ex f -> fill f (guard ex)) exs fs))

data TargetType = ForkTgt (MVar ())
                | IOTgt (IO ())

t'IOTgt :: Traversal' TargetType (IO ())
t'IOTgt k (IOTgt m) = IOTgt<$>k m
t'IOTgt _ x = return x

initCurly = do
  setLocaleEncoding utf8
  putMVar getDataFileName_ref getDataFileName
  getDataFileName "proto/vc" >>= \p -> modifyIORef vcsProtoRoots (p:)
  
ioTgt = return . IOTgt
forkTgt m = do
  v <- newEmptyMVar
  mfix $ \tid -> forkFinally (m tid) (\_ -> putMVar v ())
  return (ForkTgt v)

runTargets :: (?commandLineScripts :: [String], ?curlyConfig :: CurlyConfig, ?curlyPlex :: CurlyPlex, ?programName :: String) => [Target] -> IO ()
runTargets targetList = do
  let withParams tgt cfg = (nextParams tgt cfg,
                            let ?targetParams = cfg in do
                              let tgtHead = logLine Verbose $ format "Running target: %T" tgt
                              runTarget tgt <&> t'IOTgt %~ (tgtHead >> ))
  tgts <- sequence $ mapAccum_ withParams targetList defaultConf
  sequence_ [catch print io | IOTgt io <- tgts]
  sequence_ [takeMVar v | ForkTgt v <- tgts]

runTarget :: (?commandLineScripts :: [String], ?curlyConfig :: CurlyConfig, ?curlyPlex :: CurlyPlex, ?programName :: String,?targetParams :: TargetParams) => Target -> IO TargetType
runTarget Version = ioTgt $ putStrLn $ format "Curly, version %s. Crafted with love by Marc O. Coiffier." VERSION_curly
runTarget (Echo base x) = ioTgt $ do
  let echoLine = many' (Pure <$> many1' (noneOf "${}")
                        <+? splice)
      splice = between (several "${") (single '}') (Join <$> echoLine)
      runEchoLine (Pure s) = return s
      runEchoLine (Join l) = do
        s' <- fold <$> traverse runEchoLine l
        case words s' of
          "run":p:args -> readProcess p args ""
          "path":[p] -> return (base</>p)
          "env":[v] -> return (envVar "" v)
          "env":[v,def] -> return (envVar def v)
          cmd -> error $ format "Unrecogized command '%s'" (intercalate " " cmd)
  case matches Just echoLine x of
    Just el -> traverse runEchoLine el >>= putStrLn . fold
    Nothing -> error "Invalid echo line"
       
runTarget Help = ioTgt $ do
  let argFlags (Conditional i e arg) = i+e+argFlags arg
      argFlags (FlagDescription n _) = singleton' n
      argFlags _ = zero
      extraFlags = [(n,d) | (_,FlagDescription n d) <- ?curlyConfig]
      allFlags = foldMap (argFlags . snd) ?curlyConfig
      flagDesc | empty allFlags = "OPTION"
               | otherwise = format "(OPTION%s)" $ foldMap ("|+"+) allFlags
      indent i s = unlines (map (i+) (lines s))
  putStrLn $ format "Usage: %s%s %s..."
    ?programName (foldMap (' ':) ?commandLineScripts)
    flagDesc
  putStrLn $ indent "  " $ showOpts curlyOpts

  unless (empty extraFlags) $ do
    putStrLn "Extra options"
    sequence_ [putStrLn $ format "  +%s: %s" n d | (n,d) <- extraFlags]
    putStrLn ""
  
  let valOrigin v = c'string $ case envVar "" v of
        "" -> format " (default value, set %s to override)" v
        _ -> format " (from %s)" v
  
  putStrLn $ "Known systems: "+intercalate ", " (map show knownSystems)
  putStrLn $ format "Repositories%s:" (valOrigin "CURLY_VCS")
  VCSB_Native repos _ _ <- readIORef libraryVCS
  for_ repos $ \r -> putStrLn $ "  * "+r
  putStrLn $ format "Library cache: %s%s" curlyCacheDir (valOrigin "CURLY_LIBCACHE")
  putStrLn $ format "Server port: %p%s"   curlyPort     (valOrigin "CURLY_PORT")
  putStrLn $ format "Publisher key: %s%s" curlyPublisher (valOrigin "CURLY_PUBLISHER")
  putStr "\n"
  putStrLn $ format "Mounts:%s" (if nonempty (?curlyPlex^.mounts) then "" else " none")
  for_ (?curlyPlex^.mounts) $ putStrLn . uncurry (format "  * %s = %I" . intercalate " ")
  let showTgts = c'list $ select (/=Help) $ ?curlyPlex^.targets
  putStrLn $ format "Targets:%s" (if nonempty showTgts then "" else " none")
  traverse_ (putStrLn . format "  * %T") showTgts
runTarget Interactive = ioTgt $ runCurlySession (\_ -> unit) LocalClient =<< targetServer
runTarget (Execute s) = ioTgt $ runCurlySession (\_ -> unit) (StringClient s) =<< targetServer
runTarget (RunFile f) = ioTgt $ runCurlySession (\_ -> unit) (FileClient f) =<< targetServer
runTarget ServeInstance = forkTgt $ \tid1 -> trylog unit $ bracket_
                                             (modifyIORef processInstances (touch (getConf confInstance)))
                                             (modifyIORef processInstances (delete (getConf confInstance)))
                                             $ do
  port <- until $ (<*= maybe (threadDelay 1000000) (const unit)) $ do
    h <- peerClient
    runConnection Just True h $ do
      x <- exchange (DeclareInstance (getConf confInstance))
      case x of
        Left err -> liftIO (putStrLn err) >> zero
        Right p -> return p
  tid2 <- forkIO $ while $ catch (\e -> case fromException e of
                                    Just ThreadKilled -> return False
                                    _ -> print e >> return True)
          $ map (const True) $ do
    h <- peerClient
    runConnection_ True h $ doTimes_ 1000 $ void $ exchange (RedeclareInstance (getConf confInstance) port)
  tids <- newIORef (c'int 0,c'map zero)
  let ?targetParams = ?targetParams & confThreads %- Just (tid1,tid2,tids)
      
  putStrLn $ format "Hosting instance '%s' on port %p" (getConf confInstance) port
  sock <- listenOn port
  forever $ do
    (h,addr) <- accept sock
    logLine Verbose $ "Accepting connection from "+show addr
    n <- runAtomic tids $ l'1`swapWith`(+1)
    (`forkFinally` (\e -> case e of
                      Left exc -> logLine Debug (show exc)
                      _ -> runAtomic tids $ do l'2 =~ delete n)) $ do
      let storeTid tid = runAtomic tids $ do l'2 =~ insert n tid
      runCurlySession storeTid (SocketClient h) =<< targetServer
    
runTarget ListInstances = ioTgt $ do
  let (h,p) = case getConf confServer of
        Just (_,h,p) -> (h,p)
        Nothing -> ("127.0.0.1",curlyPort)
  h <- connectTo h p
  insts <- fold <$> runConnection Just True h (exchange AskInstances)
  putStrLn $ format "Available instances: %s" (intercalate "," insts)

runTarget (Translate f sys path) = ioTgt $ do
  runAtomic (?curlyPlex^.mountainCache) $ do l'2 =~ (doBuild:)
  withMountain $ doBuild ?mountain
  where doBuild m = let ?mountain = m in case localContext^?atMs path of
          Just (Pure (_,e)) -> do
            let prog = specializeStandalone sys e
            createDirectoryIfMissing True (dropFileName f)
            withFile f WriteMode $ \h -> writeHBytes h prog
            modifyPermissions f (_sysProgPerms sys)
          _ -> putStrLn $ "Error: the path "+show path+" doesn't seem to point to a function in the default context"
  
runTarget (Goody "builtins/ids") = ioTgt $ for_ (reverse builtinLibs) $ \l -> putStrLn (show (l^.flID))
runTarget (Goody ('b':'u':'i':'l':'t':'i':'n':'s':'/':'v':x)) = ioTgt $ do
  let (h,ext) = splitAt (length x-4) x
  case ext of
    ".cyl" -> writeHBytes stdout ((reverse builtinLibs!!(read h-1))^.flBytes)
    _ -> error $ "No such builtin library: "+x
runTarget (Goody "list") = ioTgt $ do
  fn <- getDataFileName "list"
  readBytes fn >>= writeHBytes stdout
  putStrLn "builtins/ids"
  for_ (zip [1..] builtinLibs) $ \(i,l) -> do
    putStrLn $ "builtins/v"+show i
runTarget (Goody f) = ioTgt $ do
  fn <- getDataFileName f
  readBytes fn >>= writeHBytes stdout
runTarget (SetServer _) = ioTgt unit
runTarget (SetPrelude _) = ioTgt unit
runTarget (AddPrelude _) = ioTgt unit
runTarget (SetBanner _) = ioTgt unit
runTarget (AddBanner _) = ioTgt unit
runTarget (SetInstance _) = ioTgt unit

nextParams (SetServer srv) = confServer %- srv
nextParams (SetPrelude "") = confPrelude %- []
nextParams (SetPrelude i) = confPrelude %- [i]
nextParams (AddPrelude i) = confPrelude %~ (+[i])
nextParams (SetBanner b) = confBanner %- (b+"\n")
nextParams (AddBanner b) = confBanner %~ (+b+"\n")
nextParams (SetInstance i) = confInstance %- i
nextParams _ = id

